snapcraft/internal/pluginhandler/_metadata_extraction.py

```diff
@@ -0,0 +1,62 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import importlib
+import logging
+import os
+import pkgutil
+
+from snapcraft import extractors
+from snapcraft.internal.errors import (
+    InvalidExtractorValueError,
+    MissingMetadataFileError,
+    UnhandledMetadataFileTypeError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def extract_metadata(part_name: str,
+                     file_path: str) -> extractors.ExtractedMetadata:
+    if not os.path.exists(file_path):
+        raise MissingMetadataFileError(part_name, file_path)
+
+    # Iterate through each extractor module, calling the 'extract' function
+    # from it. If it raises an 'UnhandledFileError' move onto the next.
+    for _, module_name, _ in pkgutil.iter_modules(
```

Comment:
this is clever, I am just afraid we might have some false positives
it would be nice to be able to infer some preference given the file name as well

Reply:
Yeah I was thinking about preference as well, although I couldn't think of any extractors that overlap-- do we ever anticipate that multiple extractors could handle the same file? If not, no preference is needed.